lib/remote/apilistener-configsync.cpp

```diff
@@ -462,3 +477,53 @@ void ApiListener::SendRuntimeConfigObjects(const JsonRpcConnection::Ptr& aclient
 	Log(LogInformation, "ApiListener")
 		<< "Finished syncing runtime objects to endpoint '" << endpoint->GetName() << "'.";
 }
+
+/**
+ * Wait for the given object type and name to become available for processing
+ * and lock it again before releasing the lock.
+ *
+ * @param Type::Ptr ptype The type of the object you want to wait for
+ * @param String objName The object name you want to wait for
+ */
+void ObjectNameMutex::WaitForObject(const Type::Ptr& ptype, const String& objName)
+{
+	std::unique_lock<std::mutex> lock(m_Mutex);
+	m_CV.wait(lock, [this, &ptype, &objName]{ return !IsObjectBlocked(ptype, objName); });
+
+	// Add object name to the locked list again to block all other threads that try
+	// to process a message affecting the same object.
+	m_LockedObjectNames[ptype.get()].emplace(objName);
+}
+
+/**
+ * Check whether the specified object name of the given type is in the locked list.
+ *
+ * @param Type::Ptr ptype
+ * @param String objName
+ *
+ * @return bool
+ */
+bool ObjectNameMutex::IsObjectBlocked(const Type::Ptr& ptype, const String& objName)
+{
+	auto type = m_LockedObjectNames.find(ptype.get());
+	if (type == m_LockedObjectNames.end()) {
+		return false;
+	}
```

Comment:
This is the cumbersome to check in C++ if a map/set contains an element without (accidentally) inserting it. However, if this wouldn't find the element, it would be inserted in the next step anyways, so IMHO the lock function could just do `auto& locked = m_LockedObjectNames[ptype.get()];` and then use this for all following operations, making them so simple that they don't really need a separate function.